### PR TITLE
[DO NOT SUBMIT] Template prototype

### PR DESCRIPTION
**Pull request by dvoytenko**
The initial prototype for the templating. It will be split and PR'd separately once ready.

Notes:
1. custom-template/registerTemplate added in paralled to custom-element/registerElement
2. AmpRender extension is ONLY for demonstration. Will be removed.
3. Decision matrix is here: https://docs.google.com/document/d/1UcfA2ol8q2O6WYXzZWhw8mBwyfvirIQPPBoA3sI8wvk/edit#

### CODE REVIEW

File: src/template.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+function normType(type) {
+  if (type.indexOf('text/') == 0) {
+    type = type.substring(5);
+  }
+  return type;
+}
+
+
+/**
+ * Registers an extended element. This function should typically be called
```
Comment (cramforce) [inline src/template.js]: nit: doc

File: src/template.js [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+/**
+ * Returns the promise that will eventually yield the template class. This will
+ * wait until the actual template script has been downloaded and parsed.
+ * @param {!Element} element
+ * @param {string} type
+ * @return {!Promise<!TemplateClass>}
+ */
+function waitForTemplateClass(element, type) {
+  if (knownTemplates[type]) {
```
Comment (cramforce) [inline src/template.js]: I feel you don't need this case. waitTemplates should be enough.

File: src/template.js [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    return Promise.resolve(knownTemplates[type]);
+  }
+
+  let wait = waitTemplates[type];
```
Comment (cramforce) [inline src/template.js]: This variables could just be called templateEngines.

File: test/manual/amp-render.amp.html [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  <h1>amp-render</h1>
+
+  <script type="text/amp-mustache" id="templ1">
+    Other: {{name}} at {{time}}
+    <amp-img
+        src="https://lh4.googleusercontent.com/-okOlNNHeoOc/VbYyrlFYFII/AAAAAAABYdA/La-3j3c-QQI/w600-h400-no/PANO_20150726_171347%257E2.jpg"
+        width=300 height=200 layout="responsive"></amp-img>
+  </script>
+
+  <amp-render width="300" height="300" template="templ1">
```
Comment (cramforce) [inline test/manual/amp-render.amp.html]: Reference by id and still the child?

File: third_party/mustache/README.amp [hunk 5]
```diff
@@ -0,0 +1,10 @@
+URL: https://github.com/janl/mustache.js/
+License: The MIT License
+License File: https://github.com/janl/mustache.js/blob/master/LICENSE
+
+Description:
+Fork of mustache.js library.
+
+Local Modifications:
+None yet, but expected changes for restricting instructions and sanitizing
```
Comment (cramforce) [inline third_party/mustache/README.amp]: Just for process. We want a commit that has the original sources. Ideally you have a commit that has just the third_party addition in your PR and then a second commit with all the other changes.

**Comment by cramforce** [comment]
Lets not check in .min. for mustache. We minify ourselves.

File: extensions/amp-mustache/0.1/amp-mustache.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+/** @const */
+const assert = AMP.assert;
+
+
+class AmpMustache extends AMP.BaseTemplate {
+
+  buildCallback() {
+    /** @private @const {string} */
+    this.template_ = this.element./*OK*/innerHTML;
+    mustacheParse(this.template_);
```
Comment (cramforce) [inline extensions/amp-mustache/0.1/amp-mustache.js]: Does this take significant CPU? If yes, lets do it lazily before the first render. Maybe a have a compileTemplate callback in the base class.

File: extensions/amp-render/0.1/amp-render.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    return this.element.firstChild != null;
+  }
+
+  /** @override */
+  buildCallback() {
+  }
+
+  /** @override */
+  layoutCallback() {
+    return AMP.renderTemplate(this.element, {name: 'Dima', time: new Date()}).
```
Comment (cramforce) [inline extensions/amp-render/0.1/amp-render.js]: This obviously needs data fetching.

CORS XHR. Lets make xhr.js

File: third_party/mustache/README.amp (hunk 5 above)
Comment (erwinmombay) [inline third_party/mustache/README.amp]: @cramforce question, do we want to include this in third_party, or do like register-element and install through npm and import from node_modules/* ?

Comment (cramforce) [inline third_party/mustache/README.amp]: 2 reasons for fork:

1. We expect to make changes
2. High risk of upstream inadvertently breaking security invariants.

File: src/template.js (hunk 1 above)
Comment (dvoytenko) [inline src/template.js]: Done

File: test/manual/amp-render.amp.html (hunk 4 above)
Comment (dvoytenko) [inline test/manual/amp-render.amp.html]: I was just testing two different scenarios. Done.

File: third_party/mustache/README.amp (hunk 5 above)
Comment (dvoytenko) [inline third_party/mustache/README.amp]: Will definitely do this via a separate PR + changes. And I do hope that we can just reference it via node_modules/, but I will first have to send a pull request to that project, so I hope I can first get a solid prototype going with a fork to actually figure out what might be missing in the original library. Upstream security possibility is also strong enough to warrant the fork.

File: extensions/amp-mustache/0.1/amp-mustache.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-mustache/0.1/amp-mustache.js]: I will rename this callback. It's actually what it does - lazy initialization. Poor naming.

File: extensions/amp-render/0.1/amp-render.js (hunk 7 above)
Comment (dvoytenko) [inline extensions/amp-render/0.1/amp-render.js]: I will remove this class from PR - just for testing right now. I will proceed in separate PRs with a list component and will create xhr.js. At the same time - do we not have `fetch` polyfill?

Comment (cramforce) [inline extensions/amp-render/0.1/amp-render.js]: A very limited `fetch` polyfill would be great. I believe the full spec might be way too complex.

**Comment by cramforce** [comment]
Make sure to link the decision matrix from the source code as well.

File: src/template.js (hunk 2 above)
Comment (dvoytenko) [inline src/template.js]: Done

File: src/template.js (hunk 3 above)
Comment (dvoytenko) [inline src/template.js]: Done

File: extensions/amp-mustache/0.1/amp-mustache.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-mustache/0.1/amp-mustache.js]: Done

**Comment by dvoytenko** [comment]
The prototype is switch from `<script type="text/amp-mustache">` to `<template type="amp-mustache">`. In essence, the `template` tag is polyfilled.

**Pull request closed without merging**